src/sound_openal.cpp

```diff
@@ -0,0 +1,877 @@
+	Channel* self = (Channel*)datasource;
+	int seek_offset;
+
+	switch(whence) {
+	case SEEK_CUR:
+		seek_offset = self->ogg_seekoffset + offset;
+		break;
+	case SEEK_END:
+		seek_offset = self->oggdata_lenght + offset;
+		break;
+	case SEEK_SET:
+		seek_offset = offset;
+		break;
+	/*default:
+		exit(1);*/
+	}
+	if(seek_offset > self->oggdata_lenght) return -1;
+
+	self->ogg_seekoffset = seek_offset;
+	return 0;
+}
+
+static int openal_oggclose(void* datasource) {
+	return 0;
+}
+
+static long int openal_oggtell(void* datasource) {
+	Channel* self = (Channel*)datasource;
+	return self->ogg_seekoffset;
+}
+
+static int openal_oggopen(Channel *self, const char* oggfile) {
+	FILE *f = openDataFile(oggfile, "rb");
+
+	ov_callbacks oggcb = {openal_oggread, openal_oggseek, openal_oggclose, openal_oggtell};
+
+	if(!f) {
+		return 0;
+	}
+
+	self->ogg_seekoffset = 0;
+	fseek(f, 0, SEEK_END);
+	self->oggdata_lenght = ftell(f);
+	fseek(f, 0, SEEK_SET);
+
+	self->oggdata = (char*)malloc(self->oggdata_lenght);
+	fread(self->oggdata, sizeof(char), self->oggdata_lenght, f);
+	fclose(f);
+
+	if(ov_open_callbacks(self, &self->oggStream, 0, 0, oggcb)) {
+		printf("Issues with OGG callbacks\n");
+		return 0;
+	}
+
+	self->vorbisInfo = ov_info(&self->oggStream, -1);
+	self->vorbisComment = ov_comment(&self->oggStream, -1);
+
+	alGenBuffers(4, self->streambuff);
+	return 1;
+}
+
+static int openal_oggrelease(Channel *self) {
+	alSourceStop(self->id);
+	ov_raw_seek(&self->oggStream, 0);
+	int queued;
+	alGetSourcei(self->id, AL_BUFFERS_QUEUED, &queued);
+	while(queued--) {
+		ALuint buffer;
+		alSourceUnqueueBuffers(self->id, 1, &buffer);
+	}
+	alDeleteBuffers(4, self->streambuff);
+	ov_clear(&self->oggStream);
+	free(self->oggdata);
+	return 1;
+}
+
+static int openal_streamread(Channel *self, ALuint buffer) {
+	#define OGGSIZE 65536
+	char pcm[OGGSIZE];
+	int size = 0;
+	int section;
+	int result;
+
+
+	while (size < OGGSIZE) {
+		#ifdef USE_TREMOR
+		result = ov_read(&self->oggStream, pcm+size, OGGSIZE -size, &section);
+		#else
+		result = ov_read(&self->oggStream, pcm+size, OGGSIZE -size, 0, 2, 1, &section);
+		#endif
+		if(result==0 && self->loop)
+			ov_raw_seek(&self->oggStream, 0);
+
+		if(result>0)
+			size += result;
+		else
+			break;
+	}
+
+	if(size==0) {
+		return 0;
+	}
+	alBufferData(buffer, 
+		(self->vorbisInfo->channels==1)?AL_FORMAT_MONO16:AL_FORMAT_STEREO16, 
+		pcm, size, self->vorbisInfo->rate);
+
+	return 1;
+
+	#undef OGGSIZE
+}
+
+static int openal_streamupdate(Channel* self) {
+	int processed;
+	int active = 1;
+
+	alGetSourcei(self->id, AL_BUFFERS_PROCESSED, &processed);
+
+	while(processed--) {
+		ALuint buffer;
+
+		alSourceUnqueueBuffers(self->id, 1, &buffer);
+
+		active = openal_streamread(self, buffer);
+		if(active)
+			alSourceQueueBuffers(self->id, 1, &buffer);
+	}
+	self->stream_active = active;
+
+	return active;
+}
+
+ALCcontext *openal_context = NULL;
```

Comment:
Use of `NULL` deprecated, please use `nullptr` or other things as is appropriate.